vfs/vfscache/item.go

```diff
@@ -767,6 +806,165 @@ func (item *Item) remove(reason string) (wasWriting bool) {
 	return item._remove(reason)
 }
 
+// GetCache returns the parent cache
+func (item *Item) GetCache() (cache *Cache) {
+	item.mu.Lock()
+	defer item.mu.Unlock()
+	return item.c
+}
+
+// Reset waits for cache read activities to subside, removes the cache file, then reopen it.
+// Used when cache space runs out.  Since Reset is called by the cache cleaner thread, there
+// is no direct way to return the error to the io threads.  Set item.fd to nil upon internal
+// errors so that the io threads will return internal errors seeing a nill fd.
+func (item *Item) Reset() {
+	var (
+		err error
+	)
+	item.mu.Lock()
+	defer item.mu.Unlock()
+
+	if item.opens == 0 {
+		return
+	}
+
+	item.beingReset = true
+	if item.pendingReads > 0 {
+		for {
+			if item.pendingReads == 0 {
+				break
+			}
+			item.cond.Wait()
+		}
+	}
+
+	fso := item.o
+	/*
+		item.mu.Unlock()
+		err := item.Close(nil)
+		if err != nil {
+			fs.Errorf(item.c.fcache, "vfs cache: failed to close item %s: %v", item.name, err)
+		}
+		item.remove("out of space")
+		err = item.Open(fso, true)
+		if err != nil {
+			fs.Errorf(item.c.fcache, "vfs cache: failed to reopen item %s: %v", item.name, err)
+		} */
+	// Close the downloaders
+	// Accumulate and log errors
+	checkErr := func(e error) {
+		if e != nil {
+			fs.Errorf(item.o, "vfs cache: item reset failed: %v", e)
+			if err == nil {
+				err = e
+			}
+		}
+	}
+
+	if downloaders := item.downloaders; downloaders != nil {
+		item.downloaders = nil
+		// FIXME need to unlock to kill downloader - should we
+		// re-arrange locking so this isn't necessary?  maybe
+		// downloader should use the item mutex for locking? or put a
+		// finer lock on Rs?
+		//
+		// downloader.Write calls ensure which needs the lock
+		// close downloader with mutex unlocked
+		item.mu.Unlock()
+		checkErr(downloaders.Close(nil))
+		item.mu.Lock()
+	}
+
+	// close the file handle
+	if item.fd == nil {
+		checkErr(errors.New("vfs cache item: internal error: didn't Open file"))
+	} else {
+		checkErr(item.fd.Close())
+		if err != nil {
+			// Could not close the cache file
+			return
+		}
+		item.fd = nil
+	}
+	item.mu.Unlock()
+	item.remove("cache out of space, item is clean")
+	item.mu.Lock()
+
+	item.info.ATime = time.Now()
+
+	osPath, err := item.c.mkdir(item.name) // No locking in Cache
+	if err != nil {
+		checkErr(errors.Wrap(err, "vfs cache item: open mkdir failed"))
+		return
+	}
+
+	err = item._checkObject(fso)
+	if err != nil {
+		checkErr(errors.Wrap(err, "vfs cache item: check object failed"))
+		return
+	}
+
+	// item.fd should be nil at this time
+	if item.fd != nil {
+		checkErr(errors.New("vfs cache item: internal error: didn't Close file"))
+		item.fd = nil
+		return
+	}
+	item.modified = false
+
+	fd, err := file.OpenFile(osPath, os.O_RDWR, 0600)
```

Comment:
This code is duplicated from `Open` - we should probably factor it out of there rather than repeat it...

Reply:
How about factoring the following line out from item.open() and into a new item._newCacheFile() function and call it from both item.open() and item.Reset()?

```
	item.modified = false

	fd, err := file.OpenFile(osPath, os.O_RDWR, 0600)
	if err != nil {
		item.opens--
		return errors.Wrap(err, "vfs cache item: open failed")
	}
	err = file.SetSparse(fd)
	if err != nil {
		fs.Debugf(item.name, "vfs cache: failed to set as a sparse file: %v", err)
	}
	item.fd = fd

	err = item._save()
	if err != nil {
		closeErr := item.fd.Close()
		if closeErr != nil {
			closeErr = fs.CountError(closeErr)
			log.Fatal(closeErr)
		}
		item.fd = nil
		item.opens--
		return errors.Wrap(err, "vfs cache item: _save failed")
	}
```

Reply:
Sounds good :-)